compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -2171,11 +2177,67 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                 .unwrap_or("'lt".to_string())
         };
 
-        let add_lt_sugg = generics
-            .params
-            .first()
-            .and_then(|param| param.def_id.as_local())
-            .map(|def_id| (self.tcx.def_span(def_id).shrink_to_lo(), format!("{}, ", new_lt)));
+        let mut add_lt_suggs: Vec<Option<_>> = vec![];
+        if is_synthetic {
+            if let Some(ast_generics) = ast_generics {
+                if ast_generics.params.iter().any(|p| {
+                    matches!(
+                        p.kind,
+                        hir::GenericParamKind::Lifetime { kind: hir::LifetimeParamKind::Explicit }
+                    )
+                }) {
+                    ast_generics.params.first().and_then(|param| Some(param.def_id)).map(
+                        |def_id| {
+                            add_lt_suggs.push(Some((
+                                self.tcx.def_span(def_id).shrink_to_lo(),
+                                format!("{}, ", new_lt),
+                            )));
+                        },
+                    );
+                } else {
+                    add_lt_suggs
+                        .push(Some((ast_generics.span.shrink_to_hi(), format!("<{}>", new_lt))));
+                }
+            }
+        } else {
+            generics.params.first().and_then(|param| param.def_id.as_local()).map(|def_id| {
+                add_lt_suggs.push(Some((
+                    self.tcx.def_span(def_id).shrink_to_lo(),
+                    format!("{}, ", new_lt),
+                )));
+            });
+        };
+
+        if let Some(ast_generics) = ast_generics {
+            let _ = ast_generics
+                .params
+                .iter()
+                .filter(|p| p.is_elided_lifetime())
+                .map(|p| {
+                    if self.tcx.sess.source_map()
+                        .span_to_prev_source(p.span.shrink_to_hi())
+                        .ok()
+                        .map_or(false, |s| s.as_bytes()[s.len()-1] == b'&')
+                    {
+                        add_lt_suggs
+                            .push(Some(
+                                (
+                                    p.span.shrink_to_hi(),
+                                    if let Ok(snip) = self.tcx.sess.source_map().span_to_next_source(p.span)
+                                        && snip.starts_with(' ')
```

Comment:
Can this situation be recognized without looking at the source?

Reply:
Did not find other easy ways to do this.